feat(core): implement `DependencyGraph`

arendjr opened the pull request:
## Summary

Implements a major part of #3307. Through the `DependencyGraph` we can conveniently look up which file a given import statement or expression is pointing towards. Note this only works for JS/TS files (no CSS `@import` for now).

~~Based on an unmerged PR for `oxc-resolver` (https://github.com/oxc-project/oxc-resolver/pull/360), so I'll wait with merging this until that one is merged.~~ **Update:** Bumped `oxc-resolver` to 4.0.

I major part of getting this to work with `oxc-resolver` was connecting our types for `package.json` and `tsconfig.json` with the new traits used by `oxc-resolver`. For full compatibility with `package.json` this even required storing and querying arbitrary JSON values, so I've created the `biome_json_value` crate for this, exporting types such as `JsonValue`, `JsonArray`, `JsonObject` and `JsonString`. These can be created relatively cheaply from our CST representations, while `JsonString` also contains fully unescaped string values.

###  Caveats

* The `DependencyGraph` is not yet available to lint rules, so we require one more step before making use of it.
* We still have no method of updating the `DependencyGraph` in the LSP, so information is likely to get stale when files are created/moved/deleted.

## Test Plan

Tests added, but I'm still iterating on this a bit: It's a bit of a hassle that symlinks don't work in the memory FS...

ematipico commented on crates/biome_service/src/workspace/server.rs:
```diff
@@ -221,6 +226,23 @@ impl WorkspaceServer {
             .unwrap_or(DocumentFileSource::from_path(path))
     }
 
+    fn get_js_syntax(&self, path: &Utf8Path) -> Result<AnyJsRoot, WorkspaceError> {
+        let documents = self.documents.pin();
+        let doc = documents.get(path).ok_or_else(WorkspaceError::not_found)?;
+        let file_source = self.file_sources[doc.file_source_index];
+        match file_source {
+            DocumentFileSource::Js(_) => match &doc.syntax {
+                Ok(parse) => Ok(parse.tree()),
+                Err(error) => Err(WorkspaceError::FileTooLarge(error.clone())),
+            },
+            _ => Err(WorkspaceError::source_file_not_supported(
+                file_source,
+                path.to_string(),
+                path.extension().map(str::to_string),
+            )),
+        }
+    }
```
A couple of concerns of mine:
- This ties up to one of my first comments, where potentially any language can be parsed and used for the dependency graph. So, having `get_js_syntax` isn't correct. 
- The PR doesn't explain the use of "file too large being an error", so I just assume things. Raising a `WorkspaceError::FileTooLarge` **again** feels wrong because we already do so in the method `get_parse`.

arendjr replied:
Yeah, I looked again, and I actually don't really do anything with these errors at all. Error handling is expected to be done already so I've simplified this a bit.

ematipico reviewed (commented):
I left questions to better understand the direction of the PR

arendjr commented on crates/biome_json_value/src/json_value.rs:
```diff
@@ -0,0 +1,350 @@
+    }
+}
+
+impl biome_deserialize::Deserializable for JsonValue {
```
I think I would like to challenge the idea that it's a "good practice" to put things behind feature flags in general. Feature flags add quite a bit of mental overhead for developers, because we need to look into which flags are enabled under which conditions, we need to test code both with a feature enabled and disabled, and even all the combinations that are possible among feature flags. It also slows down tooling, since toggling a feature flag causes Cargo to invalidate caches, so everything taken together they can cause quite a bit of drain in productivity.

That doesn't mean feature flags are bad, just that we should be cautious against overusing them. I can sorta see the rationale for putting the `oxc_resolver` traits behind a flag, because it's an external dependency. But for internal crates, I really wouldn't do this unless there is clear need for it. We don't "win" anything by making `biome_deserialize` optional, since it's a Biome dependency regardless and any artifacts we produce will have it included anyway. So unless there's a very specific need from users that wish to use this crate in a very specific way, I'd say we better leave it.

ematipico replied:
No worries, it was just a suggestion :)

Conaclos replied:
Makes sense, we don't put `biome_deserialize` behind a flag in other crates.

arendjr commented on crates/biome_dependency_graph/src/resolver_cache.rs:
```diff
@@ -0,0 +1,511 @@
+}
+
+impl Cache for ResolverCache<'_> {
+    type Cp = CachedPath;
+    type Pj = PackageJson;
+    type Tc = TsConfigJson;
```
Haha, yeah, that's my doing :sweat_smile: I tried to stick with what appeared to be the convention in `oxc-resolver`, but I quickly started having second thoughts on this too...

arendjr commented on crates/biome_dependency_graph/src/dependency_graph.rs:
```diff
@@ -0,0 +1,123 @@
+/// This approach makes both lookups easy and makes it very easy for us to
+/// invalidate part of the graph when there are file system changes.
+#[derive(Debug, Default)]
+pub struct DependencyGraph {
```
I intentionally didn't use `petgraph` because it's not thread-safe. It seems quite a bit harder to use as well...

ematipico replied:
Yeah it's definitely hard to use, but it's [thread-safe](https://docs.rs/petgraph/latest/petgraph/stable_graph/struct.StableGraph.html). It implements `Send` and `Sync`, unless I missed something

arendjr replied:
Ah right, I might have looked at a different type that wasn’t `Send` or `Sync`. But even then, all the mutation methods take `&mut self`, meaning we are responsible for locking in a threading environment, which is exactly what I’m trying to avoid.

arendjr commented on crates/biome_dependency_graph/src/dependency_graph.tests.rs:
```diff
@@ -0,0 +1,123 @@
+
+use super::*;
+
+fn create_test_project_layout() -> (MemoryFileSystem, ProjectLayout) {
```
I actually would like to propose doing this more often instead of always requiring a `tests/` folder. The `.tests.rs` pattern has a very nice advantage in that it works like any inline `mod tests { ... }` module, which means it can also access all the symbols from the parent module, and you don't need to make things public just for the sake of the tests.

arendjr commented on crates/biome_dependency_graph/fixtures/shared/package.json:
It's still coming :)

arendjr commented on crates/biome_package/src/node_js_package/package_json.rs:
```diff
@@ -1,30 +1,90 @@
+        }
+    }
+
+    pub fn with_version(self, version: Version) -> Self {
+        Self {
+            version: Some(version),
+            ..self
+        }
+    }
```
This is more a utility for the tests :)

arendjr commented on crates/biome_package/src/node_js_package/package_json.rs:
```diff
@@ -1,30 +1,90 @@
+        }
+    }
+
+    pub fn with_exports(self, exports: impl Into<JsonValue>) -> Self {
+        let mut raw_json = self.raw_json;
+        raw_json.insert("exports".into(), exports.into());
+        Self { raw_json, ..self }
+    }
+
+    pub fn with_dependencies(self, dependencies: Dependencies) -> Self {
```
Same, just for the tests.

arendjr commented on crates/biome_rowan/src/syntax_factory/raw_syntax.rs:
```diff
@@ -6,7 +6,7 @@ use std::marker::PhantomData;
 /// Allows third-party crates to access limited information about a `GreenNode` or construct
 /// a `GreenNode` in a limited places.
-#[derive(Debug)]
+#[derive(Clone, Debug)]
```
Good catch, could be I was just doing an experiment there and forgot to revert something. Let me check…

arendjr replied:
Indeed, can be removed.

arendjr commented on crates/biome_dependency_graph/src/import_visitor.rs:
```diff
@@ -0,0 +1,75 @@
+            AnyJsImportLike::JsModuleSource(_) => {
+                self.module_imports
+                    .static_imports
+                    .insert(specifier.to_string(), import);
```
Yes, I initially tried to do this with the syntax node, but that's what got me into a lot of thread-safety trouble, due to `SyntaxNode` not being `Sync`. But I also realized it's actually good that Rust pushed back here, since we don't even *want* the syntax node. It *should* be the string.

The reason is that the analyzer will traverse the file again anyway, but that's happening asynchronously in a worker. Giving that everything is lock-free, **if** we were to store syntax nodes here, then in theory it could happen that by the time the analyzer worker starts processing the file, a workspace worker has already inserted a newly updated version of the syntax into the dependency graph. This wouldn't happen now, since we don't ever update the graph yet, but once file watchers are in place it could happen. And then we'd be comparing different versions of the same node, which may no longer have the same position in the file. Messy.

And all that mess can actually be avoided with this simple change: The analyzer will do its own traversal anyway. It already has its own nodes, so it doesn't need a copy from the graph. So a string lookup is really all that is necessary.

And what if the running analyzer tries to query a different version of the file in the dependency graph because of the race condition I described above? Now the worst that can happen is that diagnostics from rules relying on the dependency graph cannot resolve because a *future* version of the file doesn't contain a certain import anymore. And assuming the file watcher didn't only update the dependency graph, but also triggered another re-analyze, that should be only a temporary issue and the next analyzer run will correct it already.

arendjr commented:
I've addressed most of the feedback, but I'm leaving improvements for other languages out of scope for now. I feel it's more productive now to iterate on getting JS support right, than to introduce a whole bunch of additional traits to make CSS support possible.

Conaclos commented on crates/biome_package/src/node_js_package/package_json.rs:
```diff
@@ -186,9 +382,10 @@ impl DeserializationVisitor for PackageJsonVisitor {
-                    // each package can add their own field, so we should ignore any extraneous key
-                    // and only deserialize the ones that Biome deems important
+                key => {
+                    if let Some(value) = value.into_json_value() {
+                        result.raw_json.insert(key.into(), value.into());
```
Why do we need raw JSON here?

arendjr replied:
It’s because `oxc-resolver` wants us to implement these getter functions that can pull fields out of the `package.json` using arbitrary paths.

I considered just limiting our support to a fixed set of paths, so we can pull it from pre-parsed static structures, but I tried a similar thing in the `oxc-resolver` repo when I introduced the traits, and it made the tests fail. I haven’t gone too deep into figuring out why, but I figured I just play it safe :)

Conaclos commented on crates/biome_deserialize/src/lib.rs:
```diff
@@ -142,6 +143,9 @@ pub trait DeserializableValue: Sized {
 
     /// Returns the type of this value.
     fn visitable_type(&self) -> Option<DeserializableType>;
+
+    /// Returns a JSON representation of the value, if possible.
+    fn into_json_value(self) -> Option<AnyJsonValue>;
```
I am really not fond of adding this method. This breaks a bit the overall abstraction.
I noted that the method is only used once. I don't understand why we need to handle raw json?

ematipico replied:
Neither do I, things are a bit blurry to be honest.

arendjr replied:
Removed it!

Conaclos commented:
Oh finally a dependency graph 🍬 ❤️

> For full compatibility with package.json this even required storing and querying arbitrary JSON values, so I've created the biome_json_value crate for this, exporting types such as JsonValue, JsonArray, JsonObject and JsonString. These can be created relatively cheaply from our CST representations, while JsonString also contains fully unescaped string values.

Could we directly create JsonValue, JsonArray, JsonObject and JsonString from `DeserailizableValue` (simply by implementing `Deserializable` on a sum type of these types)? This could avoid introducing the `into_json_value()` method on the `DeseraiizableValue` trait.

arendjr commented:
> Could we directly create JsonValue, JsonArray, JsonObject and JsonString from DeserailizableValue (simply by implementing Deserializable on a sum type of these types)? This could avoid introducing the into_json_value() method on the DeseraiizableValue trait.

That’s what I originally tried to use, but I couldn’t get it to work. I think it was actually based on a misunderstanding of how these traits are supposed to work. `JsonValue` already is a sum type, and `AnyJsonValue` already is too. That’s why the `into_json_value()` method works and can simply return `Some(self)`. But `JsonValue` already implements `Deserializable`, so I tried to use it as you suggest, but couldn’t get that to work with the deserialize method of `PackageJson`. Maybe I missed something?

Conaclos reviewed (commented):
> I tried to use it as you suggest, but couldn’t get that to work with the deserialize method of PackageJson. Maybe I missed something?

I made a code suggestion.
Let me know if it works and If you have any question. `biome_deserialize` is not an easy crate. I am quite satisfied of the provided abstractions. This why shy I am picky 😅

arendjr commented on crates/biome_package/src/node_js_package/package_json.rs:
```diff
@@ -186,9 +382,10 @@ impl DeserializationVisitor for PackageJsonVisitor {
-                _ => {
-                    // each package can add their own field, so we should ignore any extraneous key
-                    // and only deserialize the ones that Biome deems important
+                key => {
+                    if let Some(value) = value.into_json_value() {
+                        result.raw_json.insert(key.into(), value.into());
+                    }
```
You were totally right, thanks! I guess I just chased my tail a bit before fixing it :rofl:

arendjr commented on crates/biome_json_value/src/json_value.rs:
```diff
@@ -0,0 +1,365 @@
+///
+/// See the [module-level documentation](self) for more info.
+#[derive(Clone, Debug, Default, Deserializable, Eq, Hash, PartialEq)]
+pub struct JsonString(Text);
```
Yeah, avoiding the allocations is the reason I went for this approach...

arendjr commented on crates/biome_json_value/src/json_value.rs:
```diff
@@ -0,0 +1,365 @@
+///
+/// See the [module-level documentation](self) for more info.
+#[derive(Clone, Debug, Default, Deserializable, PartialEq)]
+pub struct JsonArray(Vec<JsonValue>);
```
I don't think that'll work, since we want these types to remain mutable.

arendjr commented on crates/biome_json_value/src/json_value.rs:
```diff
@@ -0,0 +1,365 @@
+///
+/// See the [module-level documentation](self) for more info.
+#[derive(Clone, Debug, Default, Deserializable, PartialEq)]
+pub struct JsonObject(IndexMap<JsonString, JsonValue, FxBuildHasher>);
```
What's the motivation for getting rid of `indexmap`? I'm not strongly opinionated towards it, but I think we should at least have an alternative that provides equivalent performance for lookups, which a `Vec` wouldn't provide.

Conaclos replied:
Is order preserving an important thing in our context? We could use `FxHashMap` instead.

> What's the motivation for getting rid of indexmap?

We primarily used ordered set (`indexmap::IndexSet`) in our configuration system.
However, ordered set was not a good fit in the places it was used (It created erroneous configuration merging).
Removing this usage, this leads to the use of the `indexmap` only in some places.
And, most of the time, a `Vec`, a `HashSet` or `HashMap` is sufficient.

Now we only have a few places that use the `indexmap` crate.
This raises the question of Biome's dependency on this crate.
If we only use `indexmap` two or three times in the codebase, it might be worth removing it.

arendjr replied:
I think we do rely on the order preservation here, yes. This is because the resolving of imports and exports in `package.json` iterates through the keys to find matching entries.
